File: src/services.js
'use strict';

/**
 * The four services StreamHub knows about.
 *
 * `home`        - opened when you click the service in the sidebar
 * `search`      - template used to jump to a specific title (%s = url-encoded title)
 * `tmdbNames`   - names TMDB/JustWatch may use for this provider, in order of
 *                 preference. Matched case-insensitively against the provider
 *                 list TMDB returns for your region, so we never hardcode IDs
 *                 (they change when services rebrand).
 * `deepLink`    - optional function turning a service-specific id into a direct URL
 */

const SERVICES = [
  {
    id: 'netflix',
    name: 'Netflix',
    short: 'N',
    color: '#e50914',
    accent: '#ff2d3a',
    home: 'https://www.netflix.com/browse',
    search: 'https://www.netflix.com/search?q=%s',
    tmdbNames: ['Netflix', 'Netflix basic with Ads', 'Netflix Standard with Ads'],
  },
  {
    id: 'disneyplus',
    name: 'Disney+',
    short: 'D+',
    color: '#0e2a6e',
    accent: '#2f7bff',
    home: 'https://www.disneyplus.com/home',
    search: 'https://www.disneyplus.com/search?q=%s',
    tmdbNames: ['Disney Plus', 'Disney+', 'Disney Plus Standard with Ads'],
  },
  {
    id: 'hbomax',
    name: 'HBO Max',
    short: 'HBO',
    color: '#4b1fa8',
    accent: '#8b5cf6',
    home: 'https://www.hbomax.com/',
    search: 'https://www.hbomax.com/search?q=%s',
    tmdbNames: ['HBO Max', 'Max', 'Max Amazon Channel', 'HBO Max Amazon Channel'],
  },
  {
    id: 'primevideo',
    name: 'Prime Video',
    short: 'PV',
    color: '#0f4c81',
    accent: '#00a8e1',
    home: 'https://www.primevideo.com/',
    search: 'https://www.primevideo.com/search/ref=atv_nb_sr?phrase=%s',
    tmdbNames: ['Amazon Prime Video', 'Prime Video', 'Amazon Video'],
  },
];

function byId(id) {
  return SERVICES.find((s) => s.id === id) || null;
}

/** Build the URL that jumps to a title inside a service. */
function titleUrl(serviceId, title) {
  const svc = byId(serviceId);
  if (!svc) return null;
  if (!title) return svc.home;
  return svc.search.replace('%s', encodeURIComponent(title));
}

module.exports = { SERVICES, byId, titleUrl };

File: firetv/app/src/main/java/com/felix/streamhub/data/Services.kt
package com.felix.streamhub.data

import android.graphics.Color

/**
 * The four services, plus every Android package name each of them has shipped
 * under. Names are matched against TMDB's provider list at runtime rather than
 * hardcoding provider ids, because those change whenever a service rebrands.
 */
data class Service(
    val id: String,
    val name: String,
    val short: String,
    /** Solid button background - dark enough for white text to be readable. */
    val color: Int,
    /** Brighter, used for the glow behind a service tile. */
    val accent: Int,
    val packages: List<String>,
    val tmdbNames: List<String>,
    val titleUrl: (String) -> String,
    /**
     * The service's own search, opened inside its app. Null: no search link
     * works, so the app's home screen is the best there is.
     */
    val search: Search?
)

/**
 * @property link builds the link from the title.
 * @property carriesQuery the link itself puts the title into the search.
 * @property typeQuery the link opens an empty search page whose box the
 *   accessibility service can type into (see picker/ProfilePickers.kt).
 */
data class Search(val link: (String) -> String, val carriesQuery: Boolean, val typeQuery: Boolean = false)

object Services {

    val ALL: List<Service> = listOf(
        Service(
            id = "netflix",
            name = "Netflix",
            short = "N",
            color = Color.parseColor("#c1121f"),
            accent = Color.parseColor("#e50914"),
            packages = listOf("com.netflix.ninja", "com.netflix.mediaclient"),
            tmdbNames = listOf("Netflix", "Netflix basic with Ads", "Netflix Standard with Ads"),
            titleUrl = { id -> "https://www.netflix.com/title/$id" },
            search = null
        ),
        Service(
            id = "disneyplus",
            name = "Disney+",
            short = "D+",
            color = Color.parseColor("#0e2a6e"),
            accent = Color.parseColor("#2f7bff"),
            packages = listOf("com.disney.disneyplus", "com.disney.disneyplus.androidtv"),
            tmdbNames = listOf("Disney Plus", "Disney+", "Disney Plus Standard with Ads"),
            titleUrl = { id -> "https://www.disneyplus.com/video/$id" },
            // Opens Disney+'s search page on a real Fire TV; ?q=, ?query= and
            // /search/<q> were all ignored, so the title is typed in instead.
            search = Search({ "disneyplus://www.disneyplus.com/search" }, carriesQuery = false, typeQuery = true)
        ),
        Service(
            id = "hbomax",
            name = "HBO Max",
            short = "HBO",
            color = Color.parseColor("#4b1fa8"),
            accent = Color.parseColor("#8b5cf6"),
            packages = listOf("com.wbd.stream", "com.hbo.hbonow", "com.hbo.max"),
            tmdbNames = listOf("HBO Max", "Max", "Max Amazon Channel", "HBO Max Amazon Channel"),
            titleUrl = { id -> "https://www.hbomax.com/video/watch/$id" },
            // Opens HBO Max's search page with an empty box on a real Fire TV:
            // ?q= is ignored, /search/<q> shows "Content Not Available", and its
            // screen exposes nothing to type into. You type with the remote.
            search = Search({ "https://play.max.com/search" }, carriesQuery = false)
        ),
        Service(
            id = "primevideo",
            name = "Prime Video",
            short = "PV",
            color = Color.parseColor("#0f4c81"),
            accent = Color.parseColor("#00a8e1"),
            packages = listOf("com.amazon.avod", "com.amazon.firebat", "com.amazon.avod.thirdpartyclient"),
            tmdbNames = listOf("Amazon Prime Video", "Prime Video", "Amazon Video"),
            titleUrl = { asin -> "https://www.amazon.com/gp/video/detail/$asin" },
            // Routed by com.amazon.firebat's DeepLinkRoutingActivity straight to
            // its search results, on a real Fire TV.
            search = Search({ q -> "https://app.primevideo.com/search?phrase=${enc(q)}" }, carriesQuery = true)
        )
    )

    fun byId(id: String): Service? = ALL.firstOrNull { it.id == id }

    private fun enc(s: String): String =
        java.net.URLEncoder.encode(s, "UTF-8").replace("+", "%20")
}
